File: DataStructures - Golang/implementations/data/structures/heaps/heap.go
package heaps

import "../interfaces"

type PriorityQueue struct {
  nodes []interfaces.Comparable
  size uint
}

func (heap *PriorityQueue) Insert(data interfaces.Comparable) {
  heap.nodes = append(heap.nodes, data)
  n := len(heap.nodes)
  shiftUp(n -1, heap.nodes)
  heap.size++
}

func (heap *PriorityQueue) Pop() interfaces.Comparable {
  var data interfaces.Comparable
  data, heap.nodes[0] = heap.nodes[0], heap.nodes[len(heap.nodes) - 1]
  heap.nodes = heap.nodes[:len(heap.nodes) - 1]
  shiftDown(0, heap.nodes)
  heap.size--
  return data
}

func shiftDown(i int, nodes []interfaces.Comparable) {
  n := len(nodes)
  for leftChild(i) < n {
    bigger, left := leftChild(i), leftChild(i)
    right := rightChild(i)
    if right < n {
      if nodes[left].Compare(nodes[right]) == -1 {
        bigger = right
      }
      if nodes[i].Compare(nodes[bigger]) == 1 {
        break
      }
      nodes[i], nodes[bigger] = nodes[bigger], nodes[i]
      i = bigger
    } else {
      if nodes[i].Compare(nodes[left]) == 1 {
        break
      }
      nodes[i], nodes[left] = nodes[left], nodes[i]
      i = left
    }
  }
}

func shiftUp(i int, nodes []interfaces.Comparable) {
  for i > 0 && nodes[i].Compare(nodes[parent(i)]) == 1 {
    nodes[i], nodes[parent(i)] = nodes[parent(i)], nodes[i]
    i = parent(i)
  }
}

func parent(i int) int {
  return (i - 1) / 2
}

func leftChild(i int) int {
  return 2*(i + 1) - 1
}

func rightChild(i int) int {
  return leftChild(i) + 1
}

File: book_exercises/src/main/java/datastructures/Heap.java
package datastructures;

import java.util.ArrayList;
import java.util.Collections;

public class Heap<T extends Comparable<T>> {
  private ArrayList<T> elements = new ArrayList<>();

  public void push(T data) {
    elements.add(data);
    shiftUp(elements.size() - 1);
  }

  private void shiftUp(int index) {
    while (index > 0 && elements.get(parentIndex(index)).compareTo(elements.get(index)) < 0) {
      int parent = parentIndex(index);
      Collections.swap(elements, parent, index);
      index = parent;
    }
  }

  public T pop() {
    T temp = elements.get(0);
    if (elements.size() <= 1) return temp;
    T last = elements.get(elements.size() - 1);
    elements.set(0, last);
    elements.remove(elements.size() - 1);
    shiftDown(0);
    return temp;
  }

  private void shiftDown(int index) {
    int n = elements.size();
    while (leftIndex(index) < n) {
      int right = rightIndex(index);
      int left = leftIndex(index);
      int bigger = left;
      T current = elements.get(index);
      if (right < n && elements.get(right).compareTo(elements.get(left)) > 0) bigger = right;
      if (elements.get(bigger).compareTo(current) <= 0) break;
      Collections.swap(elements, bigger, index);
      index = bigger;
    }
  }

  private int leftIndex(int n) {
    return 2 * n + 1;
  }

  private int rightIndex(int n) {
    return leftIndex(n) + 1;
  }

  private int parentIndex(int n) {
    return (n - 1) / 2;
  }
}
